tko-subs.go

```diff
@@ -23,74 +25,126 @@ import (
-	herokuusername  = flag.String("herokuusername", "", "Heroku username")
-	herokuapikey    = flag.String("herokuapikey", "", "Heroku API key")
-	herokuappname   = flag.String("herokuappname", "", "Heroku app name")
-	domain          = flag.String("domain", "", "Single domain to check")
-)
-
-//Checkiferr function as a generic check for error function
-func Checkiferr(e error) {
-	if e != nil {
-		panic(e)
-	}
-}
-
-//Info function to print pretty output
-func Info(format string, args ...interface{}) {
-	fmt.Printf("\x1b[34;1m%s\x1b[0m\n", fmt.Sprintf(format, args...))
-}
-
-//CMS struct to define the CMS data provider file
 type CMS struct {
 	Name     string `csv:"name"`
 	CName    string `csv:"cname"`
 	String   string `csv:"string"`
 	OverHTTP string `csv:"http"`
 }
 
-//tkosubsResult struct to define the results
-type tkosubsResult struct {
+type ScanResult struct {
 	Domain       string
+	Cname        string
 	Provider     string
 	IsVulnerable bool
 	IsTakenOver  bool
-	RespString   string
+	Response     string
+}
+
+type Configuration struct {
+	domainsFilePath *string
+	recordsFilePath *string
+	outputFilePath  *string
+	takeOver        *bool
+	githubtoken     *string
+	herokuusername  *string
+	herokuapikey    *string
+	herokuappname   *string
+	domain          *string
+	threadCount     *int
+}
+
+func main() {
+	config := Configuration {
+		domainsFilePath : flag.String("domains", "domains.txt", "List of domains to check"),
+		recordsFilePath : flag.String("data", "providers-data.csv", "CSV file containing CMS providers' string for identification"),
+		outputFilePath  : flag.String("output", "output.csv", "Output file to save the results"),
+		takeOver        : flag.Bool("takeover", false, "Flag to denote if a vulnerable domain needs to be taken over or not"),
+		githubtoken     : flag.String("githubtoken", "", "Github personal access token"),
+		herokuusername  : flag.String("herokuusername", "", "Heroku username"),
+		herokuapikey    : flag.String("herokuapikey", "", "Heroku API key"),
+		herokuappname   : flag.String("herokuappname", "", "Heroku app name"),
+		domain          : flag.String("domain", "", "Domains separated by ,"),
+		threadCount     : flag.Int("threads", 5, "Number of threads to run parallel")}
+	flag.Parse()
+
+	cmsRecords := loadProviders(*config.recordsFilePath)
+	var allResults []ScanResult
+
+	if *config.domain != "" {
+		for _, domain := range strings.Split(*config.domain, ",") {
+			scanResults, err := scanDomain(domain, cmsRecords, config)
+			if (err == nil) {
+				allResults = append(allResults, scanResults...)
+			} else {
+				fmt.Printf("[%s] Domain problem : %s\n", domain, err)
+			}
+		}
+	} else {
+		domainsFile, err := os.Open(*config.domainsFilePath)
+		panicOnError(err)
+		defer domainsFile.Close()
+		domainsScanner := bufio.NewScanner(domainsFile)
+
+		var wg sync.WaitGroup
+		sem := make(chan bool, *config.threadCount)
```

Comment:
can you add some comments as to why this channel is being created and used?

There are also some comments from my previous commit that would need to be changed in your new code since you changed a few function names. In other words, just make sure the comments are all changed and make sense.

Reply:
sure, on it